File: packages/csharp/src/Identifold/Identifiers.cs
using System.Globalization;
using System.Numerics;
using System.Text.RegularExpressions;

namespace Greyfoundry.Identifold;

public sealed class IdentifoldException(string code) : ArgumentException(code)
{
    public string Code { get; } = code;
}

public sealed record ParsedPublicId(string Value, string Namespace, string MachineId);
public sealed record ReferenceDefinition(string Prefix, string Profile, string Strategy, string Scope, int Width);
public sealed record NamespaceDefinition(string PublicPrefix, ReferenceDefinition? Reference);

public static partial class Identifiers
{
    private const string Data = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
    private const string Check = Data + "*~$=U";
    private const string TypeId = "0123456789abcdefghjkmnpqrstvwxyz";

    [GeneratedRegex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")]
    private static partial Regex MachinePattern();

    [GeneratedRegex("^[a-z](?:[a-z_]{0,61}[a-z])?$")]
    private static partial Regex PublicPrefixPattern();

    public static string ParseMachineId(string input)
    {
        var value = input?.ToLowerInvariant() ?? "";
        if (!MachinePattern().IsMatch(value)) throw Error("invalid_mid");
        if (value[14] != '7') throw Error("invalid_uuid_version");
        if (!"89ab".Contains(value[19])) throw Error("invalid_mid");
        return value;
    }

    public static string PublicIdFromMachineId(string machineId, string @namespace)
    {
        if (!PublicPrefixPattern().IsMatch(@namespace)) throw Error("invalid_public_prefix");
        var number = BigInteger.Parse("0" + ParseMachineId(machineId).Replace("-", ""), NumberStyles.AllowHexSpecifier);
        var suffix = new char[26];
        for (var index = suffix.Length - 1; index >= 0; index--)
        {
            number = BigInteger.DivRem(number, 32, out var remainder);
            suffix[index] = TypeId[(int)remainder];
        }
        return $"{@namespace}_{new string(suffix)}";
    }

    public static ParsedPublicId ParsePublicId(string value)
    {
        if (value != value.ToLowerInvariant()) throw Error("invalid_pid");
        var separator = value.LastIndexOf('_');
        if (separator < 0) throw Error("invalid_public_prefix");
        var @namespace = value[..separator];
        var suffix = value[(separator + 1)..];
        if (!PublicPrefixPattern().IsMatch(@namespace)) throw Error("invalid_public_prefix");
        if (suffix.Length != 26 || suffix[0] > '7') throw Error("invalid_pid");
        var number = BigInteger.Zero;
        foreach (var symbol in suffix)
        {
            var position = TypeId.IndexOf(symbol);
            if (position < 0) throw Error("invalid_pid");
            number = (number << 5) | position;
        }
        var hex = number.ToString("x32");
        var machineId = $"{hex[..8]}-{hex[8..12]}-{hex[12..16]}-{hex[16..20]}-{hex[20..]}";
        try { machineId = ParseMachineId(machineId); }
        catch (IdentifoldException) { throw Error("invalid_pid"); }
        return new ParsedPublicId(value, @namespace, machineId);
    }

    public static char CheckSymbol(string payload, bool sequential)
    {
        var alphabet = sequential ? "0123456789" : Data;
        var @base = sequential ? 10 : 32;
        var remainder = 0;
        foreach (var symbol in payload)
        {
            var position = alphabet.IndexOf(symbol);
            if (position < 0) throw Error("invalid_ref_symbol");
            remainder = (remainder * @base + position) % 37;
        }
        return Check[remainder];
    }

    public static string CreateReferenceCandidate(
        IReadOnlyList<NamespaceDefinition> registry, string @namespace, IReadOnlyList<int> randomBytes)
    {
        var reference = FindNamespace(registry, @namespace).Reference;
        if (reference is null || reference.Strategy != "random") throw Error("unknown_namespace");
        var length = ProfileLength(reference.Profile);
        if (randomBytes.Count < length) throw Error("invalid_random_source");
        var payload = new char[length];
        for (var index = 0; index < length; index++)
        {
            if (randomBytes[index] is < 0 or > 255) throw Error("invalid_random_source");
            payload[index] = Data[randomBytes[index] % 32];
        }
        var text = new string(payload);
        return $"{reference.Prefix}-{Group(text)}-{CheckSymbol(text, false)}";
    }

    public static string FormatSequentialReference(
        IReadOnlyList<NamespaceDefinition> registry, string @namespace, string sequence, string scope)
    {
        var reference = FindNamespace(registry, @namespace).Reference;
        if (reference is null || reference.Strategy != "sequence") throw Error("unknown_namespace");
        if (sequence.Length > reference.Width || sequence.Any(symbol => !char.IsAsciiDigit(symbol)))
            throw Error("sequence_overflow");
        var padded = sequence.PadLeft(reference.Width, '0');
        var payload = scope + padded;
        return $"{reference.Prefix}-{(scope.Length == 0 ? "" : scope + "-")}{padded}-{CheckSymbol(payload, true)}";
    }

    public static string Normalize(string value, IReadOnlyList<NamespaceDefinition> registry)
    {
        if (value.Contains('_'))
        {
            var parsed = ParsePublicId(value);
            FindNamespace(registry, parsed.Namespace);
            return parsed.Value;
        }
        if (value.Length == 36) return ParseMachineId(value);
        var compact = value.ToUpperInvariant().Replace("-", "");
        var definition = registry.FirstOrDefault(candidate =>
            candidate.Reference is not null && compact.StartsWith(candidate.Reference.Prefix));
        if (definition is null)
        {
            if (Regex.IsMatch(value, "^[A-Za-z]{2,8}.*")) throw Error("unknown_namespace");
            throw Error("invalid_kind");
        }
        var reference = definition.Reference!;
        var body = compact[reference.Prefix.Length..];
        if (body.Contains('?') || body.Contains('_')) throw Error("invalid_ref");
        if (reference.Strategy == "sequence")
        {
            var scopeLength = reference.Scope == "calendar-year" ? 4 : 0;
            if (body.Length != scopeLength + reference.Width + 1) throw Error("invalid_ref_length");
            var payload = body[..^1];
            if (CheckSymbol(payload, true) != body[^1]) throw Error("invalid_checksum");
            return FormatSequentialReference(registry, definition.PublicPrefix,
                payload[scopeLength..], payload[..scopeLength]);
        }
        var length = ProfileLength(reference.Profile);
        if (body.Length != length + 1) throw Error("invalid_ref_length");
        var randomPayload = body[..length].Replace('O', '0').Replace('I', '1').Replace('L', '1');
        if (CheckSymbol(randomPayload, false) != body[^1]) throw Error("invalid_checksum");
        return $"{reference.Prefix}-{Group(randomPayload)}-{body[^1]}";
    }

    private static NamespaceDefinition FindNamespace(IReadOnlyList<NamespaceDefinition> registry, string @namespace) =>
        registry.FirstOrDefault(item => item.PublicPrefix == @namespace) ?? throw Error("unknown_namespace");

    private static int ProfileLength(string profile) => profile switch
    {
        "compact" => 8,
        "high" => 12,
        _ => 10,
    };

    private static string Group(string value) => string.Join("-",
        Enumerable.Range(0, (value.Length + 3) / 4).Select(index =>
            value.Substring(index * 4, Math.Min(4, value.Length - index * 4))));

    private static IdentifoldException Error(string code) => new(code);
}

File: packages/swift/Sources/Identifold/Identifiers.swift
import Foundation

public struct ParsedPublicID: Codable, Equatable, Sendable {
  public let value: String
  public let namespace: String
  public let machineID: String

  enum CodingKeys: String, CodingKey {
    case value, namespace
    case machineID = "machineId"
  }
}

public struct ReferenceDefinition: Codable, Sendable {
  public let prefix: String
  public let profile: String?
  public let strategy: String
  public let scope: String?
  public let width: Int?
}

public struct NamespaceDefinition: Codable, Sendable {
  public let publicPrefix: String
  public let reference: ReferenceDefinition?
}

public struct IdentifoldError: Error, Sendable {
  public let code: String

  public init(_ code: String) {
    self.code = code
  }
}

public enum Identifiers {
  private static let data = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ")
  private static let check = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ*~$=U")
  private static let typeID = Array("0123456789abcdefghjkmnpqrstvwxyz")

  public static func parseMachineID(_ input: String) throws -> String {
    let value = input.lowercased()
    guard
      value.range(
        of: #"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"#,
        options: .regularExpression
      ) != nil
    else {
      throw IdentifoldError("invalid_mid")
    }
    let symbols = Array(value)
    guard symbols[14] == "7" else { throw IdentifoldError("invalid_uuid_version") }
    guard "89ab".contains(symbols[19]) else { throw IdentifoldError("invalid_mid") }
    return value
  }

  public static func publicID(from machineID: String, namespace: String) throws -> String {
    guard
      namespace.range(
        of: #"^[a-z](?:[a-z_]{0,61}[a-z])?$"#,
        options: .regularExpression
      ) != nil
    else {
      throw IdentifoldError("invalid_public_prefix")
    }
    let hex = try parseMachineID(machineID).replacingOccurrences(of: "-", with: "")
    var bits = [0, 0]
    for index in stride(from: 0, to: hex.count, by: 2) {
      let start = hex.index(hex.startIndex, offsetBy: index)
      let end = hex.index(start, offsetBy: 2)
      guard let byte = UInt8(hex[start..<end], radix: 16) else {
        throw IdentifoldError("invalid_mid")
      }
      for shift in stride(from: 7, through: 0, by: -1) {
        bits.append(Int((byte >> UInt8(shift)) & 1))
      }
    }
    var suffix = ""
    for index in stride(from: 0, to: bits.count, by: 5) {
      let value = bits[index..<(index + 5)].reduce(0) { ($0 << 1) | $1 }
      suffix.append(typeID[value])
    }
    return "\(namespace)_\(suffix)"
  }

  public static func parsePublicID(_ value: String) throws -> ParsedPublicID {
    guard value == value.lowercased() else { throw IdentifoldError("invalid_pid") }
    guard let separator = value.lastIndex(of: "_") else {
      throw IdentifoldError("invalid_public_prefix")
    }
    let namespace = String(value[..<separator])
    let suffix = String(value[value.index(after: separator)...])
    guard
      namespace.range(
        of: #"^[a-z](?:[a-z_]{0,61}[a-z])?$"#,
        options: .regularExpression
      ) != nil
    else {
      throw IdentifoldError("invalid_public_prefix")
    }
    let suffixSymbols = Array(suffix)
    guard suffixSymbols.count == 26, suffixSymbols[0] <= "7" else {
      throw IdentifoldError("invalid_pid")
    }
    var bits: [Int] = []
    for symbol in suffixSymbols {
      guard let position = typeID.firstIndex(of: symbol) else {
        throw IdentifoldError("invalid_pid")
      }
      for shift in stride(from: 4, through: 0, by: -1) {
        bits.append((position >> shift) & 1)
      }
    }
    guard bits[0] == 0, bits[1] == 0 else { throw IdentifoldError("invalid_pid") }
    bits.removeFirst(2)
    var hex = ""
    for index in stride(from: 0, to: bits.count, by: 8) {
      let byte = bits[index..<(index + 8)].reduce(0) { ($0 << 1) | $1 }
      hex += String(format: "%02x", byte)
    }
    let first = String(hex.prefix(8))
    let second = String(hex.dropFirst(8).prefix(4))
    let third = String(hex.dropFirst(12).prefix(4))
    let fourth = String(hex.dropFirst(16).prefix(4))
    let fifth = String(hex.dropFirst(20))
    let machineID = [first, second, third, fourth, fifth].joined(separator: "-")
    do {
      return ParsedPublicID(
        value: value,
        namespace: namespace,
        machineID: try parseMachineID(machineID)
      )
    } catch {
      throw IdentifoldError("invalid_pid")
    }
  }

  public static func checkSymbol(_ payload: String, sequential: Bool) throws -> Character {
    let alphabet = Array(sequential ? "0123456789" : String(data))
    let base = sequential ? 10 : 32
    var remainder = 0
    for symbol in payload {
      guard let position = alphabet.firstIndex(of: symbol) else {
        throw IdentifoldError("invalid_ref_symbol")
      }
      remainder = (remainder * base + position) % 37
    }
    return check[remainder]
  }

  public static func createReferenceCandidate(
    registry: [NamespaceDefinition],
    namespace: String,
    randomBytes: [Int]
  ) throws -> String {
    guard let reference = try findNamespace(registry, namespace).reference,
      reference.strategy == "random"
    else {
      throw IdentifoldError("unknown_namespace")
    }
    let length = profileLength(reference.profile)
    guard randomBytes.count >= length else { throw IdentifoldError("invalid_random_source") }
    var payload = ""
    for value in randomBytes.prefix(length) {
      guard (0...255).contains(value) else { throw IdentifoldError("invalid_random_source") }
      payload.append(data[value % 32])
    }
    return "\(reference.prefix)-\(group(payload))-\(try checkSymbol(payload, sequential: false))"
  }

  public static func formatSequentialReference(
    registry: [NamespaceDefinition],
    namespace: String,
    sequence: String,
    scope: String = ""
  ) throws -> String {
    guard let reference = try findNamespace(registry, namespace).reference,
      reference.strategy == "sequence"
    else {
      throw IdentifoldError("unknown_namespace")
    }
    let width = reference.width ?? 0
    guard !sequence.isEmpty,
      sequence.allSatisfy(\.isNumber),
      sequence.count <= width
    else {
      throw IdentifoldError("sequence_overflow")
    }
    let padded = String(repeating: "0", count: width - sequence.count) + sequence
    let payload = scope + padded
    let scoped = scope.isEmpty ? "" : "\(scope)-"
    return "\(reference.prefix)-\(scoped)\(padded)-\(try checkSymbol(payload, sequential: true))"
  }

  public static func normalize(_ value: String, registry: [NamespaceDefinition]) throws -> String {
    if value.contains("_") {
      let parsed = try parsePublicID(value)
      _ = try findNamespace(registry, parsed.namespace)
      return parsed.value
    }
    if value.count == 36 { return try parseMachineID(value) }

    let compact = value.uppercased().replacingOccurrences(of: "-", with: "")
    guard
      let definition = registry.first(where: {
        $0.reference.map { compact.hasPrefix($0.prefix) } ?? false
      }), let reference = definition.reference
    else {
      let letters = value.prefix { $0.isLetter }.count
      throw IdentifoldError(letters >= 2 ? "unknown_namespace" : "invalid_kind")
    }
    let body = String(compact.dropFirst(reference.prefix.count))
    guard !body.contains("?"), !body.contains("_") else {
      throw IdentifoldError("invalid_ref")
    }
    if reference.strategy == "sequence" {
      let scopeLength = reference.scope == "calendar-year" ? 4 : 0
      let width = reference.width ?? 0
      guard body.count == scopeLength + width + 1 else {
        throw IdentifoldError("invalid_ref_length")
      }
      let payload = String(body.dropLast())
      guard try checkSymbol(payload, sequential: true) == body.last else {
        throw IdentifoldError("invalid_checksum")
      }
      return try formatSequentialReference(
        registry: registry,
        namespace: definition.publicPrefix,
        sequence: String(payload.dropFirst(scopeLength)),
        scope: String(payload.prefix(scopeLength))
      )
    }
    let length = profileLength(reference.profile)
    guard body.count == length + 1 else { throw IdentifoldError("invalid_ref_length") }
    let raw = String(body.prefix(length))
    let payload = raw.map { $0 == "O" ? "0" : ($0 == "I" || $0 == "L" ? "1" : $0) }
    let normalized = String(payload)
    guard try checkSymbol(normalized, sequential: false) == body.last else {
      throw IdentifoldError("invalid_checksum")
    }
    return "\(reference.prefix)-\(group(normalized))-\(body.last!)"
  }

  private static func findNamespace(
    _ registry: [NamespaceDefinition],
    _ namespace: String
  ) throws -> NamespaceDefinition {
    guard let definition = registry.first(where: { $0.publicPrefix == namespace }) else {
      throw IdentifoldError("unknown_namespace")
    }
    return definition
  }

  private static func profileLength(_ profile: String?) -> Int {
    switch profile {
    case "compact": 8
    case "high": 12
    default: 10
    }
  }

  private static func group(_ value: String) -> String {
    let symbols = Array(value)
    return stride(from: 0, to: symbols.count, by: 4).map {
      String(symbols[$0..<min($0 + 4, symbols.count)])
    }.joined(separator: "-")
  }
}
